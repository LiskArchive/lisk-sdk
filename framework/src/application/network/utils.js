/*
 * Copyright Â© 2019 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 */

'use strict';

const net = require('net');
const dns = require('dns');

const lookupPromise = (hostname, options) =>
	new Promise((resolve, reject) => {
		dns.lookup(hostname, options, (err, address) => {
			if (err) {
				return reject(err);
			}

			return resolve(address);
		});
	});

const lookupPeersIPs = async (peersList, enabled) => {
	// If peers layer is not enabled there is no need to create the peer's list
	if (!enabled) {
		return [];
	}

	// In case domain names are used, resolve those to IP addresses.
	return Promise.all(
		peersList.map(async peer => {
			if (net.isIPv4(peer.ip)) {
				return peer;
			}

			try {
				const address = await lookupPromise(peer.ip, { family: 4 });
				return {
					...peer,
					ip: address,
				};
			} catch (err) {
				console.error(
					`Failed to resolve peer domain name ${peer.ip} to an IP address`,
				);
				return peer;
			}
		}),
	);
};

module.exports = {
	lookupPeersIPs,
};
