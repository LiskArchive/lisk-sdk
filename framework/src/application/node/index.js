/*
 * Copyright Â© 2019 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 */

'use strict';

const { config: DefaultConfig } = require('./defaults');
const Chain = require('./chain');
const { migrations } = require('./components');
const BaseModule = require('../../modules/base_module');

/* eslint-disable class-methods-use-this */

class ChainModule extends BaseModule {
	constructor(options) {
		super(options);

		this.chain = null;
	}

	static get alias() {
		return 'chain';
	}

	static get info() {
		return {
			author: 'LiskHQ',
			version: '0.1.0',
			name: 'lisk-framework-chain',
		};
	}

	static get migrations() {
		return migrations;
	}

	static get defaults() {
		return DefaultConfig;
	}

	get events() {
		return [
			'bootstrap',
			'blocks:change',
			'transactions:confirmed:change',
			'signature:change',
			'transactions:change',
			'rounds:change',
			'multisignatures:signature:change',
			'multisignatures:change',
			'delegates:fork',
			'loader:sync',
			'dapps:change',
			'rebuild',
			'processor:sync',
			'processor:deleteBlock',
			'processor:broadcast',
			'processor:newBlock',
		];
	}

	get actions() {
		return {
			calculateSupply: {
				handler: action => this.chain.actions.calculateSupply(action),
			},
			calculateMilestone: {
				handler: action => this.chain.actions.calculateMilestone(action),
			},
			calculateReward: {
				handler: action => this.chain.actions.calculateReward(action),
			},
			getForgerPublicKeysForRound: {
				handler: async action =>
					this.chain.actions.getForgerPublicKeysForRound(action),
			},
			updateForgingStatus: {
				handler: async action => this.chain.actions.updateForgingStatus(action),
			},
			postSignature: {
				handler: async action => this.chain.actions.postSignature(action),
			},
			getForgingStatusForAllDelegates: {
				handler: async () =>
					this.chain.actions.getForgingStatusForAllDelegates(),
			},
			getTransactionsFromPool: {
				handler: async action =>
					this.chain.actions.getTransactionsFromPool(action),
			},
			getTransactions: {
				handler: async action => this.chain.actions.getTransactions(action),
				isPublic: true,
			},
			getSignatures: {
				handler: async () => this.chain.actions.getSignatures(),
				isPublic: true,
			},
			postTransaction: {
				handler: async action => this.chain.actions.postTransaction(action),
			},
			getSlotNumber: {
				handler: async action => this.chain.actions.getSlotNumber(action),
			},
			calcSlotRound: {
				handler: async action => this.chain.actions.calcSlotRound(action),
			},
			getNodeStatus: {
				handler: async () => this.chain.actions.getNodeStatus(),
			},
			getLastBlock: {
				handler: async () => this.chain.actions.getLastBlock(),
				isPublic: true,
			},
			getBlocksFromId: {
				handler: async action => this.chain.actions.getBlocksFromId(action),
				isPublic: true,
			},
			getHighestCommonBlock: {
				handler: async action =>
					this.chain.actions.getHighestCommonBlock(action),
				isPublic: true,
			},
		};
	}

	async load(channel) {
		this.chain = new Chain(channel, this.options);
		await this.chain.bootstrap();
		channel.publish('chain:bootstrap');
	}

	async unload() {
		return this.chain.cleanup();
	}
}

module.exports = ChainModule;
