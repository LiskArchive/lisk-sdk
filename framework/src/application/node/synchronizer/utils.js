/*
 * Copyright Â© 2019 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 */

'use strict';

const { maxBy } = require('lodash');
const { ForkStatus } = require('@liskhq/lisk-bft');

const restoreBlocks = async (blocksModule, processorModule, tx = null) => {
	const tempBlocks = await blocksModule.dataAccess.getTempBlocks(
		{},
		{ sort: 'height:asc', limit: null },
		tx,
	);

	if (!tempBlocks || tempBlocks.length === 0) {
		return false;
	}

	for (const tempBlockEntry of tempBlocks) {
		const tempBlockInstance = await processorModule.deserialize(
			tempBlockEntry.fullBlock,
		);
		await processorModule.processValidated(tempBlockInstance, {
			removeFromTempTable: true,
		});
	}

	return true;
};

const clearBlocksTempTable = storageModule =>
	storageModule.entities.TempBlock.truncate();

const deleteBlocksAfterHeight = async (
	processorModule,
	blocksModule,
	logger,
	desiredHeight,
	backup = false,
) => {
	let { height: currentHeight } = blocksModule.lastBlock;
	logger.debug(
		{ desiredHeight, lastBlockHeight: currentHeight },
		'Deleting blocks after height',
	);
	while (desiredHeight < currentHeight) {
		logger.trace(
			{
				height: blocksModule.lastBlock.height,
				blockId: blocksModule.lastBlock.id,
			},
			'Deleting block and backing it up to temporary table',
		);
		const lastBlock = await processorModule.deleteLastBlock({
			saveTempBlock: backup,
		});
		currentHeight = lastBlock.height;
	}
};

/**
 * Allows to restore blocks if there are blocks left upon startup in temp_blocks table (e.g. node crashed)
 * Depends upon fork choice rule if blocks will be applied
 *
 * 1. Gets all blocks from temp_blocks table
 * 2. Sort blocks according to height as we want to lowest height (the next block to be applied)
 * 3. Uses next block with fork choice rule - if fork status indicates we should switch to different chain
 * we continue applying blocks using the `restoreBlocks` function.
 * Otherwise we truncate the temp_blocks table.
 */
const restoreBlocksUponStartup = async (
	logger,
	blocksModule,
	processorModule,
	storageModule,
) => {
	// Get all blocks and find lowest height (next one to be applied)
	const tempBlocks = await storageModule.entities.TempBlock.get(
		{},
		{
			sort: 'height:asc',
			limit: null,
		},
	);
	const blockLowestHeight = tempBlocks[0];
	const blockHighestHeight = tempBlocks[tempBlocks.length - 1];

	const nextTempBlock = await processorModule.deserialize(
		blockHighestHeight.fullBlock,
	);
	const forkStatus = await processorModule.forkStatus(nextTempBlock);
	const blockHasPriority =
		forkStatus === ForkStatus.DIFFERENT_CHAIN ||
		forkStatus === ForkStatus.VALID_BLOCK;

	// Block in the temp table has preference over current tip of the chain
	if (blockHasPriority) {
		logger.info('Restoring blocks from temporary table');
		await deleteBlocksAfterHeight(
			processorModule,
			blocksModule,
			logger,
			blockLowestHeight.height - 1,
			false,
		);
		// In case fork status is DIFFERENT_CHAIN - try to apply blocks from temp_blocks table
		await restoreBlocks(blocksModule, processorModule);
		logger.info('Chain successfully restored');
	} else {
		// Not different chain - Delete remaining blocks from temp_blocks table
		await clearBlocksTempTable(storageModule);
	}
};

const computeBlockHeightsList = (
	finalizedHeight,
	activeDelegates,
	listSizeLimit,
	currentRound,
) => {
	const startingHeight = Math.max(1, (currentRound - 1) * activeDelegates);
	const heightList = new Array(listSizeLimit)
		.fill(0)
		.map((_, i) => startingHeight - i * activeDelegates)
		.filter(height => height > 0);
	const heightListAfterFinalized = heightList.filter(
		height => height > finalizedHeight,
	);
	return heightList.length !== heightListAfterFinalized.length
		? [...heightListAfterFinalized, finalizedHeight]
		: heightListAfterFinalized;
};

// eslint-disable-next-line class-methods-use-this
const computeLargestSubsetMaxBy = (arrayOfObjects, propertySelectorFunc) => {
	const maximumBy = maxBy(arrayOfObjects, propertySelectorFunc);
	const absoluteMax = propertySelectorFunc(maximumBy);
	const largestSubset = [];
	// eslint-disable-next-line no-restricted-syntax
	for (const item of arrayOfObjects) {
		if (propertySelectorFunc(item) === absoluteMax) {
			largestSubset.push(item);
		}
	}
	return largestSubset;
};

module.exports = {
	computeBlockHeightsList,
	computeLargestSubsetMaxBy,
	deleteBlocksAfterHeight,
	restoreBlocksUponStartup,
	restoreBlocks,
	clearBlocksTempTable,
};
