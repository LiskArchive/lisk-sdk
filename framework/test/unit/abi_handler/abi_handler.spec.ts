/*
 * Copyright Â© 2022 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 */

import * as os from 'os';
import { StateStore, Transaction } from '@liskhq/lisk-chain';
import { codec } from '@liskhq/lisk-codec';
import { getRandomBytes } from '@liskhq/lisk-cryptography';
import { InMemoryKVStore, KVStore } from '@liskhq/lisk-db';
import { BaseModule, TokenModule } from '../../../src';
import { ABIHandler } from '../../../src/abi_handler/abi_handler';
import { transferParamsSchema } from '../../../src/modules/token/schemas';
import { StateMachine } from '../../../src/node/state_machine';
import { applicationConfigSchema } from '../../../src/schema';
import { createFakeBlockHeader } from '../../../src/testing';
import { channelMock } from '../../../src/testing/mocks';
import { genesisBlock } from '../../fixtures';
import { fakeLogger } from '../../utils/node';

describe('abi handler', () => {
	let abiHandler: ABIHandler;
	const genesis = genesisBlock();

	beforeEach(() => {
		const stateMachine = new StateMachine();
		const mod = new TokenModule();
		jest.spyOn(mod.commands[0], 'execute').mockResolvedValue();
		stateMachine.registerModule(mod as BaseModule);
		abiHandler = new ABIHandler({
			logger: fakeLogger,
			channel: channelMock,
			stateDB: (new InMemoryKVStore() as unknown) as KVStore,
			moduleDB: (new InMemoryKVStore() as unknown) as KVStore,
			genesisBlock: genesis,
			stateMachine,
			modules: [mod],
			config: applicationConfigSchema.default,
		});
	});

	describe('init', () => {
		it('should return valid response', async () => {
			jest.spyOn(os, 'homedir').mockReturnValue('/User/lisk');
			const resp = await abiHandler.init({});

			expect(resp.genesisBlock).toEqual({
				header: genesis.header.toObject(),
				assets: genesis.assets.getAll(),
				transactions: [],
			});
			expect(resp.registeredModules).toHaveLength(1);
			expect(resp.config).toMatchSnapshot();
		});
	});

	describe('initStateMachine', () => {
		it('should fail if execution context exists', async () => {
			abiHandler['_executionContext'] = {
				id: getRandomBytes(32),
			} as never;
			await expect(
				abiHandler.initStateMachine({
					header: createFakeBlockHeader().toObject(),
					networkIdentifier: getRandomBytes(32),
				}),
			).rejects.toThrow('Execution context is already initialized');
		});

		it('should create execution context and resolve context id', async () => {
			const networkIdentifier = getRandomBytes(32);
			const resp = await abiHandler.initStateMachine({
				header: createFakeBlockHeader().toObject(),
				networkIdentifier,
			});
			expect(resp.contextID).toHaveLength(32);
			expect(abiHandler['_executionContext']).not.toBeUndefined();
			expect(abiHandler['_executionContext']?.networkIdentifier).toEqual(networkIdentifier);
			expect(abiHandler['_executionContext']?.stateStore).toBeInstanceOf(StateStore);
		});
	});

	describe('initGenesisState', () => {
		it('should fail if execution context does not exist', async () => {
			await expect(
				abiHandler.initGenesisState({
					contextID: getRandomBytes(32),
					stateRoot: getRandomBytes(32),
				}),
			).rejects.toThrow('Context is not initialized or different');
		});

		it('should fail if execution context does not match', async () => {
			await abiHandler.initStateMachine({
				header: createFakeBlockHeader().toObject(),
				networkIdentifier: getRandomBytes(32),
			});
			await expect(
				abiHandler.initGenesisState({
					contextID: getRandomBytes(32),
					stateRoot: getRandomBytes(32),
				}),
			).rejects.toThrow('Context is not initialized or different');
		});

		it('should execute genesis block and resolve the response', async () => {
			jest.spyOn(abiHandler['_stateMachine'], 'executeGenesisBlock');
			const { contextID } = await abiHandler.initStateMachine({
				header: createFakeBlockHeader().toObject(),
				networkIdentifier: getRandomBytes(32),
			});
			const resp = await abiHandler.initGenesisState({
				contextID,
				stateRoot: getRandomBytes(32),
			});
			expect(abiHandler['_stateMachine'].executeGenesisBlock).toHaveBeenCalledTimes(1);

			expect(resp.events).toBeArray();
			expect(resp.assets).toEqual(genesis.assets.getAll());
			expect(resp.nextValidators).toBeArray();
			expect(resp.certificateThreshold).toEqual(BigInt(0));
			expect(resp.preCommitThreshold).toEqual(BigInt(0));
		});
	});

	describe('insertAssets', () => {
		it('should fail if execution context does not exist', async () => {
			await expect(
				abiHandler.insertAssets({
					contextID: getRandomBytes(32),
					finalizedHeight: 0,
				}),
			).rejects.toThrow('Context is not initialized or different');
		});

		it('should fail if execution context does not match', async () => {
			await abiHandler.initStateMachine({
				header: createFakeBlockHeader().toObject(),
				networkIdentifier: getRandomBytes(32),
			});
			await expect(
				abiHandler.insertAssets({
					contextID: getRandomBytes(32),
					finalizedHeight: 0,
				}),
			).rejects.toThrow('Context is not initialized or different');
		});

		it('should execute insertAssets and resolve the response', async () => {
			jest.spyOn(abiHandler['_stateMachine'], 'insertAssets');
			const { contextID } = await abiHandler.initStateMachine({
				header: createFakeBlockHeader().toObject(),
				networkIdentifier: getRandomBytes(32),
			});
			const resp = await abiHandler.insertAssets({
				contextID,
				finalizedHeight: 0,
			});
			expect(abiHandler['_stateMachine'].insertAssets).toHaveBeenCalledTimes(1);

			expect(resp.assets).toEqual([]);
		});
	});

	describe('verifyAssets', () => {
		it('should fail if execution context does not exist', async () => {
			await expect(
				abiHandler.verifyAssets({
					contextID: getRandomBytes(32),
					assets: [{ data: getRandomBytes(30), moduleID: 2 }],
				}),
			).rejects.toThrow('Context is not initialized or different');
		});

		it('should fail if execution context does not match', async () => {
			await abiHandler.initStateMachine({
				header: createFakeBlockHeader().toObject(),
				networkIdentifier: getRandomBytes(32),
			});
			await expect(
				abiHandler.verifyAssets({
					contextID: getRandomBytes(32),
					assets: [{ data: getRandomBytes(30), moduleID: 2 }],
				}),
			).rejects.toThrow('Context is not initialized or different');
		});

		it('should execute verifyAssets and resolve the response', async () => {
			jest.spyOn(abiHandler['_stateMachine'], 'verifyAssets');
			const { contextID } = await abiHandler.initStateMachine({
				header: createFakeBlockHeader().toObject(),
				networkIdentifier: getRandomBytes(32),
			});
			const resp = await abiHandler.verifyAssets({
				contextID,
				assets: [{ data: getRandomBytes(30), moduleID: 2 }],
			});
			expect(abiHandler['_stateMachine'].verifyAssets).toHaveBeenCalledTimes(1);

			expect(resp).not.toBeUndefined();
		});
	});

	describe('beforeTransactionsExecute', () => {
		it('should fail if execution context does not exist', async () => {
			await expect(
				abiHandler.beforeTransactionsExecute({
					contextID: getRandomBytes(32),
					assets: [{ data: getRandomBytes(30), moduleID: 2 }],
					consensus: {
						currentValidators: [
							{
								address: getRandomBytes(20),
								bftWeight: BigInt(1),
								blsKey: getRandomBytes(48),
								generatorKey: getRandomBytes(32),
							},
						],
						implyMaxPrevote: false,
						maxHeightCertified: 0,
					},
				}),
			).rejects.toThrow('Context is not initialized or different');
		});

		it('should fail if execution context does not match', async () => {
			await abiHandler.initStateMachine({
				header: createFakeBlockHeader().toObject(),
				networkIdentifier: getRandomBytes(32),
			});
			await expect(
				abiHandler.beforeTransactionsExecute({
					contextID: getRandomBytes(32),
					assets: [{ data: getRandomBytes(30), moduleID: 2 }],
					consensus: {
						currentValidators: [
							{
								address: getRandomBytes(20),
								bftWeight: BigInt(1),
								blsKey: getRandomBytes(48),
								generatorKey: getRandomBytes(32),
							},
						],
						implyMaxPrevote: false,
						maxHeightCertified: 0,
					},
				}),
			).rejects.toThrow('Context is not initialized or different');
		});

		it('should execute beforeTransactionsExecute and resolve the response', async () => {
			jest.spyOn(abiHandler['_stateMachine'], 'beforeExecuteBlock');
			const { contextID } = await abiHandler.initStateMachine({
				header: createFakeBlockHeader().toObject(),
				networkIdentifier: getRandomBytes(32),
			});
			const resp = await abiHandler.beforeTransactionsExecute({
				contextID,
				assets: [{ data: getRandomBytes(30), moduleID: 2 }],
				consensus: {
					currentValidators: [
						{
							address: getRandomBytes(20),
							bftWeight: BigInt(1),
							blsKey: getRandomBytes(48),
							generatorKey: getRandomBytes(32),
						},
					],
					implyMaxPrevote: false,
					maxHeightCertified: 0,
				},
			});
			expect(abiHandler['_stateMachine'].beforeExecuteBlock).toHaveBeenCalledTimes(1);

			expect(resp.events).toBeArray();
		});
	});

	describe('afterTransactionsExecute', () => {
		it('should fail if execution context does not exist', async () => {
			await expect(
				abiHandler.afterTransactionsExecute({
					contextID: getRandomBytes(32),
					assets: [{ data: getRandomBytes(30), moduleID: 2 }],
					consensus: {
						currentValidators: [
							{
								address: getRandomBytes(20),
								bftWeight: BigInt(1),
								blsKey: getRandomBytes(48),
								generatorKey: getRandomBytes(32),
							},
						],
						implyMaxPrevote: false,
						maxHeightCertified: 0,
					},
					transactions: [],
				}),
			).rejects.toThrow('Context is not initialized or different');
		});

		it('should fail if execution context does not match', async () => {
			await abiHandler.initStateMachine({
				header: createFakeBlockHeader().toObject(),
				networkIdentifier: getRandomBytes(32),
			});
			await expect(
				abiHandler.afterTransactionsExecute({
					contextID: getRandomBytes(32),
					assets: [{ data: getRandomBytes(30), moduleID: 2 }],
					consensus: {
						currentValidators: [
							{
								address: getRandomBytes(20),
								bftWeight: BigInt(1),
								blsKey: getRandomBytes(48),
								generatorKey: getRandomBytes(32),
							},
						],
						implyMaxPrevote: false,
						maxHeightCertified: 0,
					},
					transactions: [],
				}),
			).rejects.toThrow('Context is not initialized or different');
		});

		it('should execute afterTransactionsExecute and resolve the response', async () => {
			jest.spyOn(abiHandler['_stateMachine'], 'afterExecuteBlock');
			const { contextID } = await abiHandler.initStateMachine({
				header: createFakeBlockHeader().toObject(),
				networkIdentifier: getRandomBytes(32),
			});
			const resp = await abiHandler.afterTransactionsExecute({
				contextID,
				assets: [{ data: getRandomBytes(30), moduleID: 2 }],
				consensus: {
					currentValidators: [
						{
							address: getRandomBytes(20),
							bftWeight: BigInt(1),
							blsKey: getRandomBytes(48),
							generatorKey: getRandomBytes(32),
						},
					],
					implyMaxPrevote: false,
					maxHeightCertified: 0,
				},
				transactions: [],
			});
			expect(abiHandler['_stateMachine'].afterExecuteBlock).toHaveBeenCalledTimes(1);

			expect(resp.events).toBeArray();
		});
	});

	describe('verifyTransaction', () => {
		it('should execute verifyTransaction with existing context when context ID is not empty and resolve the response', async () => {
			jest.spyOn(abiHandler['_stateMachine'], 'verifyTransaction');
			const { contextID } = await abiHandler.initStateMachine({
				header: createFakeBlockHeader().toObject(),
				networkIdentifier: getRandomBytes(32),
			});
			// Add random data to check if new state store is used or not
			await abiHandler['_executionContext']?.stateStore.set(
				getRandomBytes(20),
				getRandomBytes(100),
			);
			const tx = new Transaction({
				commandID: 2,
				fee: BigInt(30),
				moduleID: 2,
				nonce: BigInt(2),
				params: getRandomBytes(100),
				senderPublicKey: getRandomBytes(32),
				signatures: [getRandomBytes(64)],
			});
			const resp = await abiHandler.verifyTransaction({
				contextID,
				networkIdentifier: getRandomBytes(32),
				transaction: tx.toObject(),
			});

			expect(abiHandler['_stateMachine'].verifyTransaction).toHaveBeenCalledTimes(1);
			expect(
				(abiHandler['_stateMachine'].verifyTransaction as jest.Mock).mock.calls[0][0][
					'_stateStore'
				],
			).toEqual(abiHandler['_executionContext']?.stateStore);
			expect(resp.result).toEqual(0);
		});

		it('should execute verifyTransaction with new context when context ID is empty and resolve the response', async () => {
			jest.spyOn(abiHandler['_stateMachine'], 'verifyTransaction');
			await abiHandler.initStateMachine({
				header: createFakeBlockHeader().toObject(),
				networkIdentifier: getRandomBytes(32),
			});
			// Add random data to check if new state store is used or not
			await abiHandler['_executionContext']?.stateStore.set(
				getRandomBytes(20),
				getRandomBytes(100),
			);
			const tx = new Transaction({
				commandID: 2,
				fee: BigInt(30),
				moduleID: 2,
				nonce: BigInt(2),
				params: getRandomBytes(100),
				senderPublicKey: getRandomBytes(32),
				signatures: [getRandomBytes(64)],
			});
			const resp = await abiHandler.verifyTransaction({
				contextID: Buffer.alloc(0),
				networkIdentifier: getRandomBytes(32),
				transaction: tx.toObject(),
			});

			expect(abiHandler['_stateMachine'].verifyTransaction).toHaveBeenCalledTimes(1);
			expect(
				(abiHandler['_stateMachine'].verifyTransaction as jest.Mock).mock.calls[0][0][
					'_stateStore'
				],
			).not.toEqual(abiHandler['_executionContext']?.stateStore);
			expect(resp.result).toEqual(0);
		});
	});

	describe('executeTransaction', () => {
		it('should execute executeTransaction with existing context when context ID is not empty and resolve the response', async () => {
			jest.spyOn(abiHandler['_stateMachine'], 'executeTransaction');
			const { contextID } = await abiHandler.initStateMachine({
				header: createFakeBlockHeader().toObject(),
				networkIdentifier: getRandomBytes(32),
			});
			// Add random data to check if new state store is used or not
			await abiHandler['_executionContext']?.stateStore.set(
				getRandomBytes(20),
				getRandomBytes(100),
			);
			const tx = new Transaction({
				commandID: 0,
				fee: BigInt(30),
				moduleID: 2,
				nonce: BigInt(2),
				params: codec.encode(transferParamsSchema, {}),
				senderPublicKey: getRandomBytes(32),
				signatures: [getRandomBytes(64)],
			});
			const resp = await abiHandler.executeTransaction({
				contextID,
				networkIdentifier: getRandomBytes(32),
				assets: [{ data: getRandomBytes(30), moduleID: 2 }],
				dryRun: false,
				header: createFakeBlockHeader().toObject(),
				transaction: tx.toObject(),
			});

			expect(abiHandler['_stateMachine'].executeTransaction).toHaveBeenCalledTimes(1);
			expect(
				(abiHandler['_stateMachine'].executeTransaction as jest.Mock).mock.calls[0][0][
					'_stateStore'
				],
			).toEqual(abiHandler['_executionContext']?.stateStore);
			expect(resp.result).toEqual(0);
		});

		it('should execute executeTransaction with new context when context ID is empty and resolve the response', async () => {
			jest.spyOn(abiHandler['_stateMachine'], 'executeTransaction');
			await abiHandler.initStateMachine({
				header: createFakeBlockHeader().toObject(),
				networkIdentifier: getRandomBytes(32),
			});
			// Add random data to check if new state store is used or not
			await abiHandler['_executionContext']?.stateStore.set(
				getRandomBytes(20),
				getRandomBytes(100),
			);
			const tx = new Transaction({
				commandID: 0,
				fee: BigInt(30),
				moduleID: 2,
				nonce: BigInt(2),
				params: codec.encode(transferParamsSchema, {}),
				senderPublicKey: getRandomBytes(32),
				signatures: [getRandomBytes(64)],
			});
			const resp = await abiHandler.executeTransaction({
				contextID: getRandomBytes(0),
				networkIdentifier: getRandomBytes(32),
				assets: [{ data: getRandomBytes(30), moduleID: 2 }],
				dryRun: true,
				header: createFakeBlockHeader().toObject(),
				transaction: tx.toObject(),
			});

			expect(abiHandler['_stateMachine'].executeTransaction).toHaveBeenCalledTimes(1);
			expect(
				(abiHandler['_stateMachine'].executeTransaction as jest.Mock).mock.calls[0][0][
					'_stateStore'
				],
			).not.toEqual(abiHandler['_executionContext']?.stateStore);
			expect(resp.result).toEqual(0);
		});
	});

	describe('commit', () => {
		it('should fail if execution context does not exist', async () => {
			await expect(
				abiHandler.commit({
					contextID: getRandomBytes(32),
					dryRun: true,
					expectedStateRoot: getRandomBytes(32),
					stateRoot: getRandomBytes(32),
				}),
			).rejects.toThrow('Context is not initialized or different');
		});

		it('should fail if execution context does not match', async () => {
			await abiHandler.initStateMachine({
				header: createFakeBlockHeader().toObject(),
				networkIdentifier: getRandomBytes(32),
			});
			await expect(
				abiHandler.commit({
					contextID: getRandomBytes(32),
					dryRun: true,
					expectedStateRoot: getRandomBytes(32),
					stateRoot: getRandomBytes(32),
				}),
			).rejects.toThrow('Context is not initialized or different');
		});

		it.todo('should resolve updated state root without saving when dryRun == true');

		it.todo(
			'should reject before saving if new state root is different from the expected stateRoot',
		);

		it.todo('should resolve updated state root after saving all the states');
	});

	describe('revert', () => {
		it('should fail if execution context does not exist', async () => {
			await expect(
				abiHandler.revert({
					contextID: getRandomBytes(32),
					expectedStateRoot: getRandomBytes(32),
					stateRoot: getRandomBytes(32),
				}),
			).rejects.toThrow('Context is not initialized or different');
		});

		it('should fail if execution context does not match', async () => {
			await abiHandler.initStateMachine({
				header: createFakeBlockHeader().toObject(),
				networkIdentifier: getRandomBytes(32),
			});
			await expect(
				abiHandler.revert({
					contextID: getRandomBytes(32),
					expectedStateRoot: getRandomBytes(32),
					stateRoot: getRandomBytes(32),
				}),
			).rejects.toThrow('Context is not initialized or different');
		});

		it.todo('should resolve updated state root');

		it.todo('should revert all the sates');
	});

	describe('clear', () => {
		it('should clear the execution context', async () => {
			await abiHandler.initStateMachine({
				header: createFakeBlockHeader().toObject(),
				networkIdentifier: getRandomBytes(32),
			});
			expect(abiHandler['_executionContext']).not.toBeUndefined();
			await abiHandler.clear({});

			expect(abiHandler['_executionContext']).toBeUndefined();
		});
	});

	describe('finalized', () => {
		it('should clean up all the finalized state', async () => {
			jest.spyOn(abiHandler['_stateDB'], 'clear');

			await abiHandler.finalize({
				finalizedHeight: 10,
			});
			expect(abiHandler['_stateDB'].clear).toHaveBeenCalledTimes(1);
		});
	});

	describe('getMetadata', () => {
		it.todo('should resolve metadata from all the modules');
	});

	describe('query', () => {
		it.todo('should query module endpoint with expected context');

		it.todo('should query plugin endpoint with expected context');
	});

	describe('prove', () => {
		it.todo('should provide proof based on the queries provided');
	});
});
