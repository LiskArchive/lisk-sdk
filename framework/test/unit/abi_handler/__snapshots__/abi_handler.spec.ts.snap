// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`abi handler init should return valid response 1`] = `
Object {
  "generator": Object {
    "force": false,
    "keys": Array [],
    "password": "",
  },
  "genesis": Object {
    "bftBatchSize": 103,
    "blockTime": 10,
    "communityIdentifier": "sdk",
    "maxTransactionsSize": 15360,
    "minFeePerByte": 1000,
  },
  "logger": Object {
    "consoleLogLevel": "none",
    "fileLogLevel": "info",
    "logFileName": "lisk.log",
  },
  "network": Object {
    "advertiseAddress": false,
    "blacklistedIPs": Array [],
    "fixedPeers": Array [],
    "hostIP": "127.0.0.1",
    "maxInboundConnections": 100,
    "maxOutboundConnections": 20,
    "port": 5000,
    "seedPeers": Array [],
    "whitelistedPeers": Array [],
  },
  "rpc": Object {
    "http": Object {
      "host": "127.0.0.1",
      "port": 8000,
    },
    "ipc": Object {
      "path": "/User/lisk/.lisk/beta-sdk-app/socket/ipc",
    },
    "modes": Array [
      "ipc",
    ],
    "ws": Object {
      "host": "127.0.0.1",
      "port": 8080,
    },
  },
  "system": Object {
    "dataPath": "/User/lisk/.lisk/beta-sdk-app",
    "keepEventsForHeights": 300,
    "maxBlockCache": 515,
    "networkVersion": "1.1",
    "version": "0.0.0",
  },
  "txpool": Object {
    "maxTransactions": 4096,
    "maxTransactionsPerAccount": 4096,
    "minEntranceFeePriority": 0n,
    "minReplacementFeeDifference": 10n,
    "transactionExpiryTime": 10800000,
  },
}
`;
